icechunk/src/ops/stats.rs

```diff
@@ -90,27 +76,59 @@ pub async fn repo_chunks_storage(
     let all_manifest_infos = all_snaps
         // this could be slightly optimized by not collecting all manifest info records into a vec
         // but we don't expect too many, and they are small anyway
-        .map_ok(|snap| stream::iter(snap.manifest_files().map(Ok).collect::<Vec<_>>()))
+        .map_ok(|snap| {
+            stream::iter(
+                snap.manifest_files().map(Ok::<_, RepositoryError>).collect::<Vec<_>>(),
+            )
+        })
         .try_flatten();
+
+    // we don't want to check manifests more than once, so we unique them by their id
     let unique_manifest_infos = try_unique_stream(|mi| mi.id.clone(), all_manifest_infos);
 
+    // we want to fetch many manifests in parallel, but not more than memory allows
+    // for this we use the StreamLimiter using the manifest size in bytes for usage
+    let limiter = &Arc::new(StreamLimiter::new(
+        max_manifest_mem_bytes.get(),
+        |m: &ManifestFileInfo| m.size_bytes as usize,
+    ));
+
+    // The StreamLimiter works by calling limit on every element before they are processed
+    let rate_limited_manifests = unique_manifest_infos
+        .and_then(|m| async move { Ok(limiter.clone().limit(m).await) });
+
     let seen_chunks = &Arc::new(Mutex::new(HashSet::new()));
     let asset_manager = &asset_manager;
 
-    let res = unique_manifest_infos
-        .map_ok(|manifest_info| async move {
-            let manifest_size = manifest_info.size_bytes;
-            manifest_chunks_storage(
-                manifest_info.id,
-                manifest_size,
-                Arc::clone(asset_manager),
-                Arc::clone(seen_chunks),
-            )
-            .await
+    let (_, res) = rate_limited_manifests
+        .map_ok(|m| async move {
+            let manifest =
+                Arc::clone(asset_manager).fetch_manifest(&m.id, m.size_bytes).await?;
+            Ok((manifest, m))
+        })
+        // Now we can buffer a bunch of fetch_manifest operations. Because we are using
+        // StreamLimiter we know memory is not going to blow up
+        .try_buffer_unordered(max_concurrent_manifest_fetches.get() as usize)
+        .map_ok(|(manifest, minfo)| async move {
+            let size = calculate_manifest_storage(manifest, Arc::clone(seen_chunks))?;
+            Ok((size, minfo))
+        })
+        // We do some more buffering to get some concurrency on the processing of the manifest file
+        // TODO: this should actually happen in a CPU bounded worker pool
```

Comment:
If you mean to spawn here I'd rather we don't

Reply:
Not exactly, but this can take some time, you are not supposed to do CPU bound shit in a tokio Task, it could block the network stack for example. But, I think this is enough for now, we can iterate.